### PR TITLE
Documentation for hooks

### PR DESCRIPTION
**Pull request by teobugslayer**

### CODE REVIEW

**Comment by teobugslayer** [comment]
ping @hdeshev

File: extending-cli.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+================= 
+
+NativeScript CLI allows you to extend its behavior and customize it for your project. 
+When the CLI executes a particular command (for example `tns build`) it checks whether you have added such extending hooks and executes them.
+Additionally, plugins can use these hooks to control the compilation of the program.
+
+What are the hooks
+==================
+
+The hooks are any executable code or node.js scripts placed in the `hooks` subdirectory of your project.
```
Comment (ErjanGavalji) [inline extending-cli.md]: I'd use "under the `hooks` subdirectory", provided there is one more level of subdirectories related to the CLI "lifecycle"

File: extending-cli.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+The hook file name must follow a strict naming scheme. The name declares:
+
+ - the action being hooked
+ - whether the user code must be called before or after the action
```
Comment (ErjanGavalji) [inline extending-cli.md]: Is this true? The TypeScript plugin puts subdirectories named `before_prepare` and `before_watch`, each containing a JavaScript file...

File: extending-cli.md (hunk 1 above)
Comment (teobugslayer) [inline extending-cli.md]: Agree

File: extending-cli.md (hunk 2 above)
Comment (teobugslayer) [inline extending-cli.md]: This is addressed later. I am not sure how to introduce both files and directories here without making the text overly complicated.

Comment (teobugslayer) [inline extending-cli.md]: I think I managed to do it somewhat.

**Comment by teobugslayer** [comment]
Ping @ikoevska

**Comment by rosen-vladimirov** [comment]
:+1:

File: extending-cli.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+```
+my-app/
+├── index.js
+├── package.json
+└── hooks/
+    ├── before-prepare.js (this is a Node.js script)
+    └── after-prepare (this is an executable file) 
+```
+
+To support multiple scripts extending the same action, you can use a different approach. Create a sud-directory in the `hooks` directory using the naming convention described above.
```
Comment (hdeshev) [inline extending-cli.md]: typo: sub-directory

File: extending-cli.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+Execute hooks in-process
+========================
+
+When your hook is a Node.js script, the CLI executes it in-process. This gives you access to the entire internal state of the CLI and all of its functions.
+The CLI assumes that this is a CommonJS module and calls its single exported function with four parameters.
+The type of the parameters are described in the .d.ts files which are part of the CLI source code [here](https://github.com/NativeScript/nativescript-cli/tree/master/lib/definitions) and [here](https://github.com/telerik/mobile-cli-lib/tree/master/definitions) 
+
+ - $logger: ILogger. Use the members of this class to show messages to the user cooperating with the CLI internal state.
+ - $projectData: IProjectData. Contains data about the project, like project directory, id, dependencies, etc.
+ - $usbLiveSyncService:ILiveSyncService. Use this service to invoke LiveSync for device or emulator.
```
Comment (hdeshev) [inline extending-cli.md]: This one can actually be used to check whether the hook runs during normal builds or during livesync. I don't know if it's important.

**Comment by hdeshev** [comment]
LGTM :+1:

**Pull request merged**